plugins/aws_ecr/lib/samson_aws_ecr/samson_plugin.rb

```diff
@@ -3,29 +3,55 @@
 module SamsonAwsEcr
   class Engine < Rails::Engine
     AMAZON_REGISTRY = /\A.*\.dkr.ecr.(?<region>[\w\-]+).amazonaws.com\z/
-    mattr_accessor :ecr_client
-    mattr_accessor(:credentials_expire_at) { Time.current }
 
-    initializer :ecr_client do
-      if match = AMAZON_REGISTRY.match(ENV['DOCKER_REGISTRY'])
-        SamsonAwsEcr::Engine.ecr_client = Aws::ECR::Client.new(region: match['region'])
+    class << self
+      # we make sure the repo exists so pushes do not fail
+      # - ignores if the repo already exists
+      # - ignores if the repo cannot be created due to permission problems
+      def ensure_repository(repository)
+        return unless ecr_client
+        name = repository.split('/', 2).last
+        ecr_client.describe_repositories(repository_names: [name])
+      rescue Aws::ECR::Errors::RepositoryNotFoundException
+        ecr_client.create_repository(repository_name: name)
```

Comment:
it is possible, but in both cases it will end up in a AccessDenied error
which is caught below

added docs for needed permissions

On Tue, Jun 14, 2016 at 3:51 PM, rata <notifications@github.com> wrote:

> In plugins/aws_ecr/lib/samson_aws_ecr/samson_plugin.rb
> <https://github.com/zendesk/samson/pull/1065#discussion_r67072460>:
>
> >
> > -    initializer :ecr_client do
> > -      if match = AMAZON_REGISTRY.match(ENV['DOCKER_REGISTRY'])
> > -        SamsonAwsEcr::Engine.ecr_client = Aws::ECR::Client.new(region: match['region'])
> > +    class << self
> > +      # we make sure the repo exists so pushes do not fail
> > +      # - ignores if the repo already exists
> > +      # - ignores if the repo cannot be created due to permission problems
> > +      def ensure_repository(repository)
> > +        return unless ecr_client
> > +        name = repository.split('/', 2).last
> > +        ecr_client.describe_repositories(repository_names: [name])
> > +      rescue Aws::ECR::Errors::RepositoryNotFoundException
> > +        ecr_client.create_repository(repository_name: name)
>
> Is it possible to have permissions to describe but not create? What will
> happen in that case? Or we can just document that the credentials should
> have permission to create?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1065/files/d38d9861ce655fc9cbf6f8f2d1538221e46fc7d2#r67072460>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAsZ2y1ZsdjE0J0OAXOnuM3Rlf07B1zks5qLzCGgaJpZM4I0y0W>
> .
>